packages/auth/src/grant/routes.ts

```diff
@@ -193,36 +228,82 @@ async function finishInteraction(
-  const { clientNonce, interactNonce, interactRef } = grant
-  const interactUrl = config.identityServerDomain + `/interact/${interactId}`
+  ctx.body = {
+    redirectUri:
+      config.authServerDomain +
+      `/${grant.interactId}/${grant.interactNonce}/finish`
+  }
+}
 
-  // https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol#section-4.2.3
-  const data = `${clientNonce}\n${interactNonce}\n${interactRef}\n${interactUrl}`
+async function finishInteraction(
+  deps: ServiceDependencies,
+  ctx: AppContext
+): Promise<void> {
+  const { id: interactId, nonce } = ctx.params
+  const sessionNonce = ctx.session.nonce
+
+  // TODO: redirect with this error in query string
+  if (sessionNonce !== nonce) {
+    ctx.status = 401
+    ctx.body = {
+      error: 'invalid_request'
+    }
+    return
+  }
+
+  const { grantService, config } = deps
+  const grant = await grantService.getByInteractionSession(interactId, nonce)
 
-  const hash = crypto.createHash('sha3-512').update(data).digest('base64')
-  clientRedirectUri.searchParams.set('hash', hash)
-  clientRedirectUri.searchParams.set('interact_ref', interactRef)
-  ctx.redirect(clientRedirectUri.toString())
+  // TODO: redirect with this error in query string
+  if (!grant) {
+    ctx.status = 404
+    ctx.body = {
+      error: 'unknown_request'
+    }
+    return
+  }
+
+  const clientRedirectUri = new URL(grant.finishUri)
+  if (grant.state === GrantState.Granted) {
+    const { clientNonce, interactNonce, interactRef } = grant
+    const interactUrl = config.identityServerDomain + `/interact/${interactId}`
+
+    // https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol#section-4.2.3
+    const data = `${clientNonce}\n${interactNonce}\n${interactRef}\n${interactUrl}`
+
+    const hash = crypto.createHash('sha3-512').update(data).digest('base64')
+    clientRedirectUri.searchParams.set('hash', hash)
+    clientRedirectUri.searchParams.set('interact_ref', interactRef)
+    ctx.redirect(clientRedirectUri.toString())
+  } else if (grant.state === GrantState.Rejected) {
+    clientRedirectUri.searchParams.set('result', 'grant_rejected')
+    ctx.redirect(clientRedirectUri.toString())
+  } else {
+    // Grant is not in either an accepted or rejected state
+    clientRedirectUri.searchParams.set('result', 'grant_invalid')
+    ctx.redirect(clientRedirectUri.toString())
+  }
 }
 
-async function denyInteraction(
+// TODO: allow idp to specify the reason for rejection
+async function rejectGrant(
   deps: ServiceDependencies,
   ctx: AppContext
 ): Promise<void> {
-  const { interactId } = ctx.params
-  const interactSession = ctx.session.interactId
+  // TODO: check redis for a session
+  const { id: interactId, nonce } = ctx.params
+
+  const { grantService, config } = deps
 
-  if (!interactSession || !interactId || interactSession !== interactId) {
+  if (ctx.headers['x-idp-secret'] !== deps.config.identityServerSecret) {
```

Comment:
small nit: if we compare secrets like this we should be done [safe timing compares](https://nodejs.org/api/crypto.html#cryptotimingsafeequala-b)